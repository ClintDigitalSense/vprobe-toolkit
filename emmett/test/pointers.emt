typedef struct {
   uint8  byte;
   uint16 word;
   uint32 dword;
   uint64 qword;
} testa;

typedef struct {
   testa *aptr1;
   testa *aptr2;
} testb;

typedef struct {
   testa aNest;
   testb bNest;
} testc;

testa *ptr;
testb *ptr2;
testc *ptr3;

VMM1Hz {
   ptr = GSBASE;
   ptr3 = KERNELGSBASE;
   ptr2 = ptr3->bNest.aptr2 & 0xfff;
   printf("%d\n", (ptr->byte + ptr->word + ptr->dword - ptr->qword) / 
          ptr2->aptr2->dword);
   printf("%d\n", ptr3->aNest.qword * ptr3->bNest.aptr2->byte);
   printf("%d\n", (&(ptr3->aNest))->qword == ptr3->aNest.qword);
}
