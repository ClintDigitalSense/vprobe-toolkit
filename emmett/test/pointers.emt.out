(version 0.2)
(definteger ptr)
(definteger ptr2)
(definteger ptr3)

 ;;;;;
(vprobe VMM1Hz ( do  ( setint ptr   GSBASE)( setint ptr3   KERNELGSBASE)( setint ptr2    (& (getguest  (+  (+ ptr3  16 )  8 ))  4095 ))( printf    "%d\n"   (/  (-  (+  (+ (& 255 (getguest ptr))  (& 65535 (getguest  (+ ptr  2 ))) )  (& 4294967295 (getguest  (+ ptr  4 ))) )  (getguest  (+ ptr  8 )) )  (& 4294967295 (getguest  (+ (getguest  (+ ptr2  8 ))  4 ))) ) )( printf    "%d\n"   (* (getguest  (+ ptr3  8 ))  (& 255 (getguest (getguest  (+  (+ ptr3  16 )  8 )))) ) )( printf    "%d\n"   (== (getguest  (+ ptr3  8 ))  (getguest  (+ ptr3  8 )) ) )))